### PR TITLE
add parallel-execution-framework.md

### LINKED ISSUE #92: Write down the first version of Execution Framework section

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What issue does this PR solve?

- close #92
- to #xxxx

### CODE REVIEW

**Review by tisonkun — COMMENTED** [review]
Thanks for preparing this PR @guo-shaoge ! Please add blanks between paragraphs.

File: src/understand-tidb/execution-framework.md [hunk 1]
```diff
@@ -1 +1,41 @@
 # Execution Framework
```
Comment (guo-shaoge) [inline src/understand-tidb/execution-framework.md:1]: Maybe we should use `Parallelism Execution` instead of `Execution Framework`.

Comment (tisonkun) [reply src/understand-tidb/execution-framework.md:1]: This section is named by @XuHuaiyu , what do you want to express at this section?

Comment (XuHuaiyu) [reply src/understand-tidb/execution-framework.md:1]: 1. We can rename it as `Parallel Execution Framework`.
2. We'd better add some examples or figures to make the description easier to be undertood. e.g. The difference of the implementation between intra-parallelism and inter-parallelism.

File: src/understand-tidb/execution-framework.md [hunk 2]
```diff
@@ -1 +1,41 @@
...
+
+In order to make full use of the multi-core ability of modern hardware, most popular DBMS have implemented the ability of parallel execution SQL execution engine.
+
+There are three common parallel implementations:
+1. Intra operator parallelism.
+2. Exchange operator.
+3. Morsel-Driven parallelism.
```
Comment (tisonkun) [inline src/understand-tidb/execution-framework.md:8]:
```suggestion
There are three common parallel implementations: intra-operator parallelism, exchange operator, and Morsel-Driven parallelism.
```

As you have just three sections for each.

File: src/understand-tidb/execution-framework.md [hunk 3]
```diff
@@ -1 +1,41 @@
...
+
+## Intra-operator Parallelism
+
+In intra-operator parallelism, multiple goroutines will be created inside the operator for parallel processing. The creation, termination and synchronization of the goroutines are handled by the operator itself.For example, in HashJoin, after the hashtable on the build side is built, multiple goroutines will lookup the hashtable at the same time to speed up the execution of HashJoin.And the parallel implementation of each operator is different. Hashagg is divided into two stages. Each stage will start multiple goroutines.It's still the traditional Volcano-model, and the `Next()` is still called by a single thread. The parallelism only occurs in the internal processing of the operator.
```
Comment (tisonkun) [inline src/understand-tidb/execution-framework.md:12]:
```suggestion
In intra-operator parallelism, multiple goroutines will be created inside the operator for parallel processing. The creation, termination and synchronization of the goroutines are handled by the operator itself.

For example, in HashJoin, after the hashtable on the build side is built, multiple goroutines will lookup the hashtable at the same time to speed up the execution of HashJoin. 

The parallel implementation of operators is various. HashAgg is divided into two stages. Each stage will start multiple goroutines. It still follows the traditional Volcano-model, and the `Next()` method is still called by a single thread. The parallelism only occurs in the internal processing of the operator.
```

File: src/understand-tidb/execution-framework.md [hunk 4]
```diff
@@ -1 +1,41 @@
...
+
+The degree of parallelism(DOP) can be controlled by the session variable. For example, in TiDB `tidb_executor_Concurrency` is 5 by default. it means HashJoin will create 5 goroutines to probe HashTable when it's opened.You can also control the parallelism of a specific operator by setting the session variable, such as `tidb_hash_join_concurrency`.
```
Comment (tisonkun) [inline src/understand-tidb/execution-framework.md:14]:
```suggestion
The degree of parallelism (DOP) can be controlled by the session variable. For example, `tidb_executor_concurrency` is 5 by default. It means HashJoin will create five goroutines to probe HashTable when it's opened. You can also control the parallelism of a specific operator by setting the session variable, such as `tidb_hash_join_concurrency`.
```

File: src/understand-tidb/execution-framework.md [hunk 5]
```diff
@@ -1 +1,41 @@
...
+
+## Exchange Operator
+
+Intra operator parallelism is the most intuitive way to implement parallelism, but the implementation is more complex, because every operator needs to implement parallelism independently.Moreover, too many threads will increase the scheduling overhead. However, the use of goroutines can alleviate this problem.
+
+The most traditional way to implement parallelism is to use the exchange operator. By encapsulating the parallel logic into the exchange operator, all other operators only need to consider the single thread implementation.The exchange operator is responsible for data exchange between different threads. The internal implementation of exchange is divided into two parts: sender and receiver.And data is transferred from sender to receiver in different ways, such as hash partition, random distribution or sort merge.
```
Comment (tisonkun) [inline src/understand-tidb/execution-framework.md:20]:
```suggestion
A more traditional way to implement parallelism is to use the exchange operator. By encapsulating the parallel logic into the exchange operator, all other operators only need to consider the single thread implementation.

The exchange operator is responsible for data exchange between different threads. The internal implementation of exchange is divided into two parts: sender and receiver. And data is transferred from sender to receiver in different ways, such as hash partition, random distribution or sort merge.
```

File: src/understand-tidb/execution-framework.md [hunk 6]
```diff
@@ -1 +1,41 @@
...
+
+There are usually two ways to control DOP:
```
Comment (tisonkun) [inline src/understand-tidb/execution-framework.md:22]:
```suggestion
There are usually two ways to control DOP:

```

File: src/understand-tidb/execution-framework.md [hunk 7]
```diff
@@ -1 +1,41 @@
...
+1. Users can use hints to specify DOP explicitly.
+2. The optimizer can choose DOP according to the table size of the scan operation automatically.
+
+This approach requires the optimizer to generate parallel plans. Generally, the generation of plans is divided into two stages. The first stage generates serial plans, and the second stage generates its corresponding parallel plans. The second stage is mainly responsible for inserting the exchange operator into the plan tree at the appropriate position. Both heuristic rules and costs model can be used to get the optimal parallel plan
```
Comment (tisonkun) [inline src/understand-tidb/execution-framework.md:26]:
```suggestion
The latter approach requires the optimizer to generate parallel plans. Generally, plan generation is divided into two stages. The first stage generates serial plans, and the second stage generates its corresponding parallel plans. The second stage is mainly responsible for inserting the exchange operator into the plan tree at the appropriate position. Both heuristic rules and costs model can be used to get the optimal parallel plan.
```

**Review by tisonkun — COMMENTED** [review]
Thanks for your contribution @guo-shaoge !

Comments inline. And one comment here:

As TiDB doesn't implement "Morsel-Driven parallelism", I'd prefer you mention it before break down into sub sections. And the final section can be a summary of content above.

**Review by tisonkun — COMMENTED** [review]
Generally looks good. Please also update SUMMARY.md with the new title of "Execution Framework".

/cc @zhangyangyu @xxchan 

You may be interested to be the first audience of this section.

**Comment by guo-shaoge** [comment]
> Generally looks good. Please also update SUMMARY.md with the new title of "Execution Framework".
> 
> /cc @zhangyangyu @xxchan
> 
> You may be interested to be the first audience of this section.

Done~

**Review by tisonkun — APPROVED** [review]
LGTM.

Wait a while for potential reviews. I'll merge this PR tomorrow if no other objections.

**Comment by tisonkun** [comment]
Merging...

**Pull request merged**